src/js/libcs/Render2D.js

```diff
@@ -499,60 +535,68 @@ Render2D.drawpoint = function(pt) {
     csctx.stroke();
 };
 
-Render2D.drawline = function(homog) {
-    var na = CSNumber.abs(homog.value[0]).value.real;
-    var nb = CSNumber.abs(homog.value[1]).value.real;
-    var nc = CSNumber.abs(homog.value[2]).value.real;
-    var divi;
-
-    if (na >= nb && na >= nc) {
-        divi = homog.value[0];
-    }
-    if (nb >= na && nb >= nc) {
-        divi = homog.value[1];
-    }
-    if (nc >= nb && nc >= na) {
-        divi = homog.value[2];
-    }
-    var a = CSNumber.div(homog.value[0], divi);
-    var b = CSNumber.div(homog.value[1], divi);
-    var c = CSNumber.div(homog.value[2], divi); //TODO Realitycheck einbauen
-
-    var l = [
-        a.value.real,
-        b.value.real,
-        c.value.real
-    ];
-    var b1, b2;
-    if (Math.abs(l[0]) < Math.abs(l[1])) {
-        b1 = [1, 0, 30];
-        b2 = [-1, 0, 30];
-    } else {
-        b1 = [0, 1, 30];
-        b2 = [0, -1, 30];
-    }
-    var erg1 = [
-        l[1] * b1[2] - l[2] * b1[1],
-        l[2] * b1[0] - l[0] * b1[2],
-        l[0] * b1[1] - l[1] * b1[0]
-    ];
-    var erg2 = [
-        l[1] * b2[2] - l[2] * b2[1],
-        l[2] * b2[0] - l[0] * b2[2],
-        l[0] * b2[1] - l[1] * b2[0]
-    ];
-
-    var pt1 = {
-        x: erg1[0] / erg1[2],
-        y: erg1[1] / erg1[2]
+Render2D.clipLineCore = function(a, b, c) {
+    // clip to canvas boundary (up to line size)
+    var margin = Math.sqrt(0.5) * Render2D.lsize;
+    var xMin = 0 - margin;
+    var xMax = csw + margin;
+    var yMax = 0 - margin;
+    var yMin = csh + margin;
+    var distNeg = function(x, y) {
+        return x * a + y * b + c < 0;
     };
-    var pt2 = {
-        x: erg2[0] / erg2[2],
-        y: erg2[1] / erg2[2]
+    var ul = distNeg(xMin, yMax);
+    var ur = distNeg(xMax, yMax);
+    var ll = distNeg(xMin, yMin);
+    var lr = distNeg(xMax, yMin);
+    var res = [];
+    if (ul !== ur) res.push({
+        x: (-c - b * yMax) / a,
+        y: yMax
+    });
+    if (ur !== lr) res.push({
+        x: xMax,
+        y: (-c - a * xMax) / b
+    });
+    if (ll !== lr) res.push({
+        x: (-c - b * yMin) / a,
+        y: yMin
+    });
+    if (ul !== ll) res.push({
+        x: xMin,
+        y: (-c - a * xMin) / b
+    });
+
+    return res;
+};
 
-    };
+Render2D.clipLine = function(homog) {
+    // transformation to canvas coordinates
+    var m = csport.drawingstate.matrix;
+    var mInverse = List.inverse(List.realMatrix([
+        [m.a, -m.b, m.tx],
+        [m.c, -m.d, -m.ty],
+        [0, 0, 1]
+    ]));
+    var n = List.normalizeMax(List.productMV(List.transpose(mInverse), homog));
```

Comment:
Compare this to [my `csport.toMat`](https://github.com/gagern/CindyJS/blob/7198d9d0f76080528c91006a9d05022badf6408a/src/js/libgeo/GeoState.js#L87-L95) from #280. Perhaps you want to include that here? If so, please make sure to use the same source code location, so that git will recognize the two additions as equivalent and not require manual conflict resolution.